File: 004_glfw_go/src/qtnn.go
package main

import (
	"fmt"
	"math"
)

type qtnn [4]float32

func qtnnShow(q qtnn) {
	fmt.Printf("%5.2f %5.2f %5.2f %5.2f\n", q[_XC], q[_YC], q[_ZC], q[_WC])
}

func qtnnLenght(q qtnn) float32 {
	return float32(math.Sqrt(float64(
		q[_XC]*q[_XC] +
			q[_YC]*q[_YC] +
			q[_ZC]*q[_ZC] +
			q[_WC]*q[_WC])))
}

func qtnnNormalize(q qtnn) (rt qtnn) {
	var (
		len float32
	)

	len = qtnnLenght(q)

	if len != 0.0 {
		rt[_WC] = q[_WC] / len
		rt[_XC] = q[_XC] / len
		rt[_YC] = q[_YC] / len
		rt[_ZC] = q[_ZC] / len
	}

	return rt
}

func qtnnInverse(q qtnn) (rt qtnn) {
	rt[_WC] = q[_WC]
	rt[_XC] = -q[_XC]
	rt[_YC] = -q[_YC]
	rt[_ZC] = -q[_ZC]

	return rt
}

func qtnnScale(q qtnn, scale float32) (rt qtnn) {
	rt[_WC] = q[_WC] * scale
	rt[_XC] = q[_XC] * scale
	rt[_YC] = q[_YC] * scale
	rt[_ZC] = q[_ZC] * scale
	return rt
}

func qtnnSum(a, b qtnn) (rt qtnn) {
	rt[0] = a[0] + b[0]
	rt[1] = a[1] + b[1]
	rt[2] = a[2] + b[2]
	rt[3] = a[3] + b[3]

	return rt
}

func qtnnSub(a, b qtnn) (rt qtnn) {
	rt[0] = a[0] - b[0]
	rt[1] = a[1] - b[1]
	rt[2] = a[2] - b[2]
	rt[3] = a[3] - b[3]

	return rt
}

func qtnnDot(a, b qtnn) float32 {
	return a[0]*b[0] + a[1]*b[1] + a[2]*b[2] + a[3]*b[3]
}

func qtnnMult(a, b qtnn) (rt qtnn) {
	rt[_WC] = a[_WC]*b[_WC] - a[_XC]*b[_XC] - a[_YC]*b[_YC] - a[_ZC]*b[_ZC]
	rt[_XC] = a[_WC]*b[_XC] + a[_XC]*b[_WC] + a[_YC]*b[_ZC] - a[_ZC]*b[_YC]
	rt[_YC] = a[_WC]*b[_YC] - a[_XC]*b[_ZC] + a[_YC]*b[_WC] + a[_ZC]*b[_XC]
	rt[_ZC] = a[_WC]*b[_ZC] + a[_XC]*b[_YC] - a[_YC]*b[_XC] + a[_ZC]*b[_WC]

	return rt
}

/* Не работает */
func qtnnMultVec3(a qtnn, b vec3) (rt qtnn) {
	rt[_WC] = -a[_WC]*b[_XC] - a[_YC]*b[_YC] - a[_ZC]*b[_ZC]
	rt[_XC] = a[_WC]*b[_XC] + a[_YC]*b[_ZC] - a[_ZC]*b[_YC]
	rt[_YC] = a[_WC]*b[_YC] - a[_XC]*b[_ZC] + a[_ZC]*b[_XC]
	rt[_ZC] = a[_WC]*b[_ZC] + a[_XC]*b[_YC] - a[_YC]*b[_XC]

	return rt
}

func qtnnFromVec3(v vec3) (rt qtnn) {
	rt[_XC] = v[_XC]
	rt[_YC] = v[_YC]
	rt[_ZC] = v[_ZC]
	rt[_WC] = 0.0

	return rt
}

func qtnnFromAxisAngl(a vec3, phi float64) (rt qtnn) {
	var (
		sinhalfphi float32
	)

	sinhalfphi = float32(math.Sin(phi * 0.5))

	rt[_WC] = float32(math.Cos(phi * 0.5))
	rt[_XC] = a[_XC] * sinhalfphi
	rt[_YC] = a[_YC] * sinhalfphi
	rt[_ZC] = a[_ZC] * sinhalfphi

	return rt
}

func qtnnFromEuler(yaw, pitch, roll float64) (rt qtnn) {
	var (
		qyaw, qpitch, qroll qtnn
	)

	qyaw = qtnnFromAxisAngl(vec3Set(1.0, 0.0, 0.0), yaw)
	qpitch = qtnnFromAxisAngl(vec3Set(0.0, 1.0, 0.0), pitch)
	qroll = qtnnFromAxisAngl(vec3Set(0.0, 0.0, 1.0), roll)

	rt = qtnnMult(qyaw, qpitch)

	rt = qtnnMult(rt, qroll)

	return rt
}

func qtnnToVec3(q qtnn) (rt vec3) {
	return vec3Set(q[_XC], q[_YC], q[_ZC])
}

func qtnnTransformVec3(a qtnn, b vec3) (rt vec3) {
	var (
		vq, tmp qtnn
	)

	vq = qtnnFromVec3(b)

	tmp = qtnnMult(a, vq)
	tmp = qtnnMult(tmp, qtnnInverse(a))

	return qtnnToVec3(tmp)
}

File: internal/algebra/qtnn.cpp

#include <cmath>
#include <iostream>

#include "qtnn.h"

qtnn::qtnn(float yaw, float pitch, float roll) {
    qtnn qyaw, qpitch, qroll, rt;
    vec3 vyaw, vpitch, vroll;

    vyaw = vec3(1.0, 0.0, 0.0);
    vpitch = vec3(0.0, 1.0, 0.0);
    vroll = vec3(0.0, 0.0, 1.0);

    qyaw = qtnn(vyaw, yaw);
    qpitch = qtnn(vpitch, pitch);
    qroll = qtnn(vroll, roll);

    rt = qtnnMult(qyaw, qpitch);

    rt = qtnnMult(rt, qroll);

    data[3] = rt[3];
    data[0] = rt[0];
    data[1] = rt[1];
    data[2] = rt[2];
}

qtnn qtnnScale(const qtnn &q, float scale) {
    qtnn rt;

    rt[3] = q[3] * scale;
    rt[0] = q[0] * scale;
    rt[1] = q[1] * scale;
    rt[2] = q[2] * scale;

    return rt;
}

qtnn qtnnSum(const qtnn &a, const qtnn &b) {
    qtnn rt;

    rt[0] = a[0] + b[0];
    rt[1] = a[1] + b[1];
    rt[2] = a[2] + b[2];
    rt[3] = a[3] + b[3];

    return rt;
}

qtnn qtnnSub(const qtnn &a, const qtnn &b) {
    qtnn rt;

    rt[0] = a[0] - b[0];
    rt[1] = a[1] - b[1];
    rt[2] = a[2] - b[2];
    rt[3] = a[3] - b[3];

    return rt;
}

float qtnnDot(const qtnn &a, const qtnn &b) {
    return a[0] * b[0] + a[1] * b[1] + a[2] * b[2] + a[3] * b[3];
}

qtnn qtnnMult(const qtnn &a, const qtnn &b) {
    qtnn rt;

    rt[3] = a[3] * b[3] - a[0] * b[0] - a[1] * b[1] - a[2] * b[2];
    rt[0] = a[3] * b[0] + a[0] * b[3] + a[1] * b[2] - a[2] * b[1];
    rt[1] = a[3] * b[1] - a[0] * b[2] + a[1] * b[3] + a[2] * b[0];
    rt[2] = a[3] * b[2] + a[0] * b[1] - a[1] * b[0] + a[2] * b[3];

    return rt;
}

qtnn qtnnSlerp(qtnn from, qtnn to, float t) {
    qtnn rt;
    float p1[4];
    float omega, cosom, sinom, scale0, scale1;

    // косинус угла
    cosom = qtnnDot(from, to);

    if (cosom < 0.0) {
        cosom = -cosom;
        p1[0] = -to[0];
        p1[1] = -to[1];
        p1[2] = -to[2];
        p1[3] = -to[3];
    } else {
        p1[0] = to[0];
        p1[1] = to[1];
        p1[2] = to[2];
        p1[3] = to[3];
    }

    if ((1.0 - cosom) > f_eps) {
        // стандартный случай (slerp)
        omega = acos(cosom);
        sinom = sinf(omega);
        scale0 = sinf((1.0 - t) * omega) / sinom;
        scale1 = sinf(t * omega) / sinom;
    } else {
        // если маленький угол - линейная интерполяция
        scale0 = 1.0 - t;
        scale1 = t;
    }

    rt[0] = scale0 * from[0] + scale1 * p1[0];
    rt[1] = scale0 * from[1] + scale1 * p1[1];
    rt[2] = scale0 * from[2] + scale1 * p1[2];
    rt[3] = scale0 * from[3] + scale1 * p1[3];

    return rt;
}
